### ISSUE TITLE
[aarch64] `nn.Linear(20, 1)` inference fails

### ISSUE THREAD

**Issue by pvardanis** [pytorch/pytorch#114750]
### 🐛 Describe the bug

Hi,

As the title indicates, I have a model created with `TorchScript` and saved with the `.pt` extension under `torch==1.13.1` but when I load it back in another env using `torch==2.1.1` it fails to infer.

The error is the following:
```
I'm also having errors when running a pt fail saved with TorchScript that was implemented with torch==1.13.1, but torch==2.1.1 raises the following:

249, in forward
        x = self.fc1(x)
        x = torch.nn.functional.relu(x)
        x = self.fc2(x)
            ~~~~~~~~ <--- HERE
        return x
  File "/Users/panagiotisvardanis/miniconda3/envs/mac-pytorch/lib/python3.8/site-packages/torch/nn/modules/linear.py", line 114, in forward
    def forward(self, input: Tensor) -> Tensor:
        return F.linear(input, self.weight, self.bias)
               ~~~~~~~~ <--- HERE
RuntimeError: could not create a primitive descriptor for a matmul primitive
```

### Versions

Pip requirements from the environment where the model was created:
```
[pip3] torch==1.13.1
[pip3] torchvision==0.14.0
```

Pip requirements from the environment where the model is failing to infer:
```
[pip3] torch==2.1.1
[pip3] torchvision==0.16.1
```

cc @ezyang @gchanan @zou3519 @kadeng @EikanWang @jgong5 @wenzhe-nrv @sanchitintel @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @mingfeima @ashokei @jingxu10 @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @malfet @snadampal

**Comment by malfet** [comment]
@pvardanis can you please provide a full reproducer? (Or share a .pt script if it's not sensitive)
And please run `python3 -mtorch.utils.collect_env` (I.e. looks like it happens on MacOS, but is it x86 or M1 Mac machine?)

**Comment by pvardanis** [comment]
Here's the file: [single_io_model.pt.zip](https://github.com/pytorch/pytorch/files/13502460/single_io_model.pt.zip)

I've run this but I forgot to copy paste the CPU part. It's an Apple M1 Max.

**Comment by malfet** [comment]
@pvardanis just to confirm: you are running it under inside the docker on M1?
For my own curiosity, why this appeals to you? (Not an expert, but docker on Mac is not very has a non-negligible overhead compared to running it natively, isn't it?)

**Comment by pvardanis** [comment]
@malfet It runs as part of a k8s cluster, inside a docker container for a company product. We have different `torch` wheels depending on the platform and architecture. We're developing the codebase under M1 processors but the product engine runs in different setups, e.g. `arm64`, `aarch64`, `x86_64` etc. On the particular error case, I'm running under `aarch64` and `torch==2.1.1`.

**Comment by malfet** [comment]
One line reproducer:
```
python -c "import torch;m=torch.nn.Linear(20, 1);print(torch.__version__);print(m(torch.rand(1, 20)))"
```

**Comment by malfet** [comment]
Ok, looks like try-catch is a safe solution for 2.1.2, but this is something that should eventually be fixed in oneDNN

**Comment by malfet** [comment]
Ok, looks like it is fixed by https://github.com/pytorch/pytorch/pull/114851 which is already targeted for 2.1.2, so removing the milestone from this issue, but want to add generic try catch in time for 2.2.0

**Comment by milpuz01** [comment]
Hi @malfet,

Thank you very much for following up and looking at the fix.

> Ok, looks like try-catch is a safe solution for 2.1.2, but this is something that should eventually be fixed in oneDNN

I think here we cannot fix this in oneDNN as it is doing the right thing so try-catch at call site should be the right solution.

Another way to reproduce this issue without applying fix from #110150  is with this piece of code:
```
import torch
dim0 = 32
dim1 = 1
l = torch.nn.Linear(dim0, dim1)
l.eval()
l(torch.randn((dim1, dim0)))
```

If you run this you will get error because at this call site: https://github.com/pytorch/pytorch/blob/a7bcc78bff65876d3e9e03e17fb2f86121591a8f/aten/src/ATen/native/LinearAlgebra.cpp#L1503 dimensions of LHS are 32x1, dimensions of RHS 1x32 and results is 1x1 so when creating `matmul` primitive this check: https://github.com/oneapi-src/oneDNN/blob/c8c94afd313a75eadeef66a63f5f0ddfe2d80a30/src/common/matmul.cpp#L139 is going to fail before first dimension in LHS (src) is 32 and the first dimension in output (dst) is 1.

If you change `dim1` in above code to 2 then we will not see failure because then LHS would 2x32, RHS would be 32x2 and output would be 2x2 so checks in oneDNN will pass and in that case `transpose_c` is set to `true`. Another alternative to make this to work for case when `transpose_c` is `false` is not to swap `a` and `b` in call site to `mkldnn_matmul`.

Hope this helps.
